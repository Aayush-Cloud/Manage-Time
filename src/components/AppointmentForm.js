// @flow
import React from 'react';
import { Keyboard, View } from 'react-native';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import { Field, Fields } from 'redux-form';
import i18n from 'react-native-i18n';
import moment from 'moment';

/* eslint-disable import/no-unresolved, import/extensions */
import DatePickerCollapsible from './DatePickerCollapsible';
import Textfield from './Textfield';
import TimePickerCollapsible from './TimePickerCollapsible'; // uses platform specific code
/* eslint-enable import/no-unresolved, import/extensions */
import styles from './styles/AppointmentForm.styles';
import type { Collapsible } from './CollapsibleField';

type Props = {
  onRecurrencePress: (input: { value: any }) => void,
};

type FieldProps = {
  input: {
    onPress: () => void,
    value: any,
  },
};

type FieldRef = {
  getRenderedComponent: () => Collapsible,
};

function RecurrenceRow({ input, onRecurrencePress }: Props & FieldProps) {
  return (
    <Textfield
      isButton
      onPress={() => onRecurrencePress(input)}
      editable={false}
      value={i18n.t(input.value)}
      label={i18n.t('repeat')}
    />
  );
}

export default class AppointmentForm extends React.Component<Props, {}> {
  _fieldRefs = {};

  _closeExpandedField = () => {
    if (this._lastExpandedField) {
      this._lastExpandedField.getRenderedComponent().collapse();
    }
    this._lastExpandedField = null;
  };

  _handleExpand = (fieldRef: FieldRef | null) => {
    if (fieldRef !== null) {
      if (this._lastExpandedField !== fieldRef) {
        this._closeExpandedField();
      }
      Keyboard.dismiss();
      this._lastExpandedField = fieldRef;
    }
  };

  _lastExpandedField: ?FieldRef;
  _fieldRefs: {
    timeRef: FieldRef | null,
    startdateRef: FieldRef | null,
    enddateRef: FieldRef | null,
  };

  render() {
    const { onRecurrencePress } = this.props;
    const { timeRef, startdateRef, enddateRef } = this._fieldRefs;
    return (
      <KeyboardAwareScrollView
        style={styles.container}
        keyboardShouldPersistTaps="handled"
      >
        <View style={styles.navBarSpacer} />
        <Fields
          withRef
          ref={field => {
            this._fieldRefs.timeRef = field;
          }}
          names={['starttime', 'endtime']}
          component={TimePickerCollapsible}
          label={i18n.t('time')}
          parse={(value: Date) => value && moment(value)}
          format={(value: moment) => value && value.toDate()}
          onExpand={() => this._handleExpand(timeRef)}
        />
        <View style={styles.spacer} />
        <Field
          withRef
          ref={field => {
            this._fieldRefs.startdateRef = field;
          }}
          name={'startdate'}
          component={DatePickerCollapsible}
          label={i18n.t('starts')}
          onExpand={() => this._handleExpand(startdateRef)}
        />
        <View style={styles.spacer} />
        <Field
          withRef
          ref={field => {
            this._fieldRefs.enddateRef = field;
          }}
          name={'enddate'}
          component={DatePickerCollapsible}
          label={i18n.t('ends')}
          onExpand={() => this._handleExpand(enddateRef)}
        />
        <View style={styles.spacer} />
        <Field
          name={'recurrence'}
          component={RecurrenceRow}
          onRecurrencePress={onRecurrencePress}
        />
        <View style={styles.spacer} />
        <View style={styles.spacer} />
        <Field
          name={'location'}
          component={Textfield}
          placeholder={i18n.t('location')}
          onFocus={this._closeExpandedField}
        />
        <View style={styles.spacer} />
        <Field
          name={'type'}
          component={Textfield}
          placeholder={i18n.t('type')}
          onFocus={this._closeExpandedField}
        />
      </KeyboardAwareScrollView>
    );
  }
}
