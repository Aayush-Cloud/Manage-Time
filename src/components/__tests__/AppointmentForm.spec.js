// @flow
import React from 'react';
import { shallow } from 'enzyme';
import AppointmentForm from '../AppointmentForm';

const props = {
  onRecurrencePress: jest.fn(),
};

describe('AppointmentForm component', () => {
  it('renders input fields for appointment data', () => {
    const wrapper = shallow(<AppointmentForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
