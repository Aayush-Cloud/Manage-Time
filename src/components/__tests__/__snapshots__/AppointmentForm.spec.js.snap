// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentForm component renders input fields for appointment data 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <KeyboardAwareScrollView
    enableAutoAutomaticScroll={true}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="handled"
    style={
        Object {
            "backgroundColor": "white",
            "paddingHorizontal": 16,
          }
    }
>
    <View
        style={
            Object {
                "height": 32,
              }
        }
    />
    <Fields
        component={[Function]}
        format={[Function]}
        label="Time"
        names={
            Array [
                "starttime",
                "endtime",
              ]
        }
        onExpand={[Function]}
        parse={[Function]}
        withRef={true}
    />
    <View
        style={
            Object {
                "height": 24,
              }
        }
    />
    <Field
        component={[Function]}
        label="Starts"
        name="startdate"
        onExpand={[Function]}
        withRef={true}
    />
    <View
        style={
            Object {
                "height": 24,
              }
        }
    />
    <Field
        component={[Function]}
        label="Ends"
        name="enddate"
        onExpand={[Function]}
        withRef={true}
    />
    <View
        style={
            Object {
                "height": 24,
              }
        }
    />
    <Field
        component={[Function]}
        name="recurrence"
        onRecurrencePress={[Function]}
    />
    <View
        style={
            Object {
                "height": 24,
              }
        }
    />
    <View
        style={
            Object {
                "height": 24,
              }
        }
    />
    <Field
        component={[Function]}
        name="location"
        onFocus={[Function]}
        placeholder="Location"
    />
    <View
        style={
            Object {
                "height": 24,
              }
        }
    />
    <Field
        component={[Function]}
        name="type"
        onFocus={[Function]}
        placeholder="Type (e.g. lecture, tutorial)"
    />
</KeyboardAwareScrollView>,
  "nodes": Array [
    <KeyboardAwareScrollView
      enableAutoAutomaticScroll={true}
      enableResetScrollToCoords={true}
      extraHeight={75}
      extraScrollHeight={0}
      keyboardOpeningTime={250}
      keyboardShouldPersistTaps="handled"
      style={
            Object {
                  "backgroundColor": "white",
                  "paddingHorizontal": 16,
                }
      }
>
      <View
            style={
                  Object {
                        "height": 32,
                      }
            }
      />
      <Fields
            component={[Function]}
            format={[Function]}
            label="Time"
            names={
                  Array [
                        "starttime",
                        "endtime",
                      ]
            }
            onExpand={[Function]}
            parse={[Function]}
            withRef={true}
      />
      <View
            style={
                  Object {
                        "height": 24,
                      }
            }
      />
      <Field
            component={[Function]}
            label="Starts"
            name="startdate"
            onExpand={[Function]}
            withRef={true}
      />
      <View
            style={
                  Object {
                        "height": 24,
                      }
            }
      />
      <Field
            component={[Function]}
            label="Ends"
            name="enddate"
            onExpand={[Function]}
            withRef={true}
      />
      <View
            style={
                  Object {
                        "height": 24,
                      }
            }
      />
      <Field
            component={[Function]}
            name="recurrence"
            onRecurrencePress={[Function]}
      />
      <View
            style={
                  Object {
                        "height": 24,
                      }
            }
      />
      <View
            style={
                  Object {
                        "height": 24,
                      }
            }
      />
      <Field
            component={[Function]}
            name="location"
            onFocus={[Function]}
            placeholder="Location"
      />
      <View
            style={
                  Object {
                        "height": 24,
                      }
            }
      />
      <Field
            component={[Function]}
            name="type"
            onFocus={[Function]}
            placeholder="Type (e.g. lecture, tutorial)"
      />
</KeyboardAwareScrollView>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <AppointmentForm
        onRecurrencePress={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": AppointmentForm {
        "_closeExpandedField": [Function],
        "_fieldRefs": Object {},
        "_handleExpand": [Function],
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "onRecurrencePress": [Function],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <KeyboardAwareScrollView
          enableAutoAutomaticScroll={true}
          enableResetScrollToCoords={true}
          extraHeight={75}
          extraScrollHeight={0}
          keyboardOpeningTime={250}
          keyboardShouldPersistTaps="handled"
          style={
                    Object {
                              "backgroundColor": "white",
                              "paddingHorizontal": 16,
                            }
          }
>
          <View
                    style={
                              Object {
                                        "height": 32,
                                      }
                    }
          />
          <Fields
                    component={[Function]}
                    format={[Function]}
                    label="Time"
                    names={
                              Array [
                                        "starttime",
                                        "endtime",
                                      ]
                    }
                    onExpand={[Function]}
                    parse={[Function]}
                    withRef={true}
          />
          <View
                    style={
                              Object {
                                        "height": 24,
                                      }
                    }
          />
          <Field
                    component={[Function]}
                    label="Starts"
                    name="startdate"
                    onExpand={[Function]}
                    withRef={true}
          />
          <View
                    style={
                              Object {
                                        "height": 24,
                                      }
                    }
          />
          <Field
                    component={[Function]}
                    label="Ends"
                    name="enddate"
                    onExpand={[Function]}
                    withRef={true}
          />
          <View
                    style={
                              Object {
                                        "height": 24,
                                      }
                    }
          />
          <Field
                    component={[Function]}
                    name="recurrence"
                    onRecurrencePress={[Function]}
          />
          <View
                    style={
                              Object {
                                        "height": 24,
                                      }
                    }
          />
          <View
                    style={
                              Object {
                                        "height": 24,
                                      }
                    }
          />
          <Field
                    component={[Function]}
                    name="location"
                    onFocus={[Function]}
                    placeholder="Location"
          />
          <View
                    style={
                              Object {
                                        "height": 24,
                                      }
                    }
          />
          <Field
                    component={[Function]}
                    name="type"
                    onFocus={[Function]}
                    placeholder="Type (e.g. lecture, tutorial)"
          />
</KeyboardAwareScrollView>,
        "_debugID": 2,
        "_renderedOutput": <KeyboardAwareScrollView
          enableAutoAutomaticScroll={true}
          enableResetScrollToCoords={true}
          extraHeight={75}
          extraScrollHeight={0}
          keyboardOpeningTime={250}
          keyboardShouldPersistTaps="handled"
          style={
                    Object {
                              "backgroundColor": "white",
                              "paddingHorizontal": 16,
                            }
          }
>
          <View
                    style={
                              Object {
                                        "height": 32,
                                      }
                    }
          />
          <Fields
                    component={[Function]}
                    format={[Function]}
                    label="Time"
                    names={
                              Array [
                                        "starttime",
                                        "endtime",
                                      ]
                    }
                    onExpand={[Function]}
                    parse={[Function]}
                    withRef={true}
          />
          <View
                    style={
                              Object {
                                        "height": 24,
                                      }
                    }
          />
          <Field
                    component={[Function]}
                    label="Starts"
                    name="startdate"
                    onExpand={[Function]}
                    withRef={true}
          />
          <View
                    style={
                              Object {
                                        "height": 24,
                                      }
                    }
          />
          <Field
                    component={[Function]}
                    label="Ends"
                    name="enddate"
                    onExpand={[Function]}
                    withRef={true}
          />
          <View
                    style={
                              Object {
                                        "height": 24,
                                      }
                    }
          />
          <Field
                    component={[Function]}
                    name="recurrence"
                    onRecurrencePress={[Function]}
          />
          <View
                    style={
                              Object {
                                        "height": 24,
                                      }
                    }
          />
          <View
                    style={
                              Object {
                                        "height": 24,
                                      }
                    }
          />
          <Field
                    component={[Function]}
                    name="location"
                    onFocus={[Function]}
                    placeholder="Location"
          />
          <View
                    style={
                              Object {
                                        "height": 24,
                                      }
                    }
          />
          <Field
                    component={[Function]}
                    name="type"
                    onFocus={[Function]}
                    placeholder="Type (e.g. lecture, tutorial)"
          />
</KeyboardAwareScrollView>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <AppointmentForm
    onRecurrencePress={[Function]}
/>,
}
`;
